packages/clerk-js/src/ui/elements/Select.tsx

```diff
@@ -0,0 +1,316 @@
+import React, { useState } from 'react';
+
+import { usePopover, useSearchInput } from '../../ui/hooks';
+import { Button, Flex, Icon, Text } from '../customizables';
+import { Caret, MagnifyingGlass } from '../icons';
+import { animations, PropsOfComponent } from '../styledSystem';
+import { colors } from '../utils';
+import { InputWithIcon } from './InputWithIcon';
+
+type SelectSearchOptions = {
+  placeholder: string;
+};
+
+type OptionBuilder<T> = (option: T, index?: number, isSelected?: boolean) => JSX.Element;
+
+type SearchableOption<T> = T & { searchTerm: string };
+
+type DefaultOption = SearchableOption<Record<string, unknown>> | string | number;
+
+type SelectProps<T> = PropsOfComponent<typeof Flex> & {
+  options: T[];
+  optionBuilder?: OptionBuilder<T>;
+  shownOptionBuilder?: (option: T) => JSX.Element;
+  searchOptions?: SelectSearchOptions;
+  noResultsMessage?: string;
+  defaultIndex?: number;
+  isDisabled?: boolean;
+  handleSelect?: (option: T) => void;
+};
+
+const defaultOptionBuilder = <T extends DefaultOption>(option: T, _index?: number, isSelected?: boolean) => {
+  return (
+    <Flex
+      sx={theme => ({
+        width: '100%',
+        paddingLeft: theme.space.$2,
+        paddingRight: theme.space.$2,
+        ...(isSelected && { backgroundColor: theme.colors.$blackAlpha200 }),
+        '&:hover': {
+          backgroundColor: theme.colors.$blackAlpha200,
+        },
+      })}
+    >
+      {['string', 'number'].includes(typeof option) && <Text>{option}</Text>}
+    </Flex>
+  );
+};
+
+export const Select = <T extends DefaultOption>(props: SelectProps<T>) => {
+  const {
+    isDisabled,
+    handleSelect,
+    options,
+    searchOptions,
+    noResultsMessage,
+    optionBuilder = defaultOptionBuilder,
+    shownOptionBuilder,
+    defaultIndex = 0,
+    sx,
+    ...rest
+  } = props;
+  const { floating, reference, styles, toggle: togglePopover, isOpen } = usePopover({ autoUpdate: false });
+  const [selectedIndex, setSelectedIndex] = useState(defaultIndex);
+  const selectedItemRef = React.useRef<HTMLDivElement>(null);
+  const { filteredItems, searchInputProps } = useSearchInput({
+    items: options,
+    comparator: (term, item) => {
+      if (typeof item === 'object') {
+        return item.searchTerm.toLowerCase().includes(term.toLowerCase());
```

Comment:
❓ What's the type of item? It's safer to check for `item.searchTerm` instead of checking if it's an object.

Reply:
If it's an object(and not a string or number) it will have the searchTerm property based on the types.